File: java/leetcode_grind/Day319.java
package leetcode_grind;

public class Day319 {
    // https://leetcode.com/problems/decoded-string-at-index/description/
    static class Solution {
        public String decodeAtIndex(String s, int k) {
            long size = 0;
            var n = s.length();

            for (var i = 0; i < n; i++) {
                var c = s.charAt(i);
                if (Character.isDigit(c)) {
                    size *= c - '0';
                } else {
                    size++;
                }
            }

            for (var i = n - 1; i >= 0; i--) {
                var c = s.charAt(i);
                k %= size;
                if (k == 0 && Character.isLetter(c)) {
                    return Character.toString(c);
                }

                if (Character.isDigit(c)) {
                    size /= c - '0';
                } else {
                    size--;
                }
            }

            return "";
        }
    }
}

File: src/leetcode_grind/day_319.rs
// https://leetcode.com/problems/decoded-string-at-index
pub fn decode_at_index(s: String, k: i32) -> String {
    let mut k = k as i64;
    let mut size = 0;
    for ch in s.chars() {
        if ch.is_digit(10) {
            size *= ch.to_digit(10).unwrap() as i64;
        } else {
            size += 1;
        }
    }

    for ch in s.chars().rev() {
        k %= size;
        if k == 0 && ch.is_alphabetic() {
            return format!("{}", ch);
        }
        if ch.is_alphabetic() {
            size -= 1;
        } else {
            size /= ch.to_digit(10).unwrap() as i64;
        }
    }
    "".to_string()
}
